File: app/src/main/java/com/thanksmister/iot/mqtt/alarmpanel/network/DarkSkyOptions.java
package com.thanksmister.iot.mqtt.alarmpanel.network;

import android.text.TextUtils;

import dpreference.DPreference;

/**
 * For original implementation see https://github.com/androidthings/sensorhub-cloud-iot.
 */
public class DarkSkyOptions {
    
    public static final String PREF_WEATHER_UNITS = "pref_weather_units";
    public static final String PREF_DARK_SKY_KEY = "pref_dark_sky_key";
    public static final String PREF_WEATHER_LAT = "pref_weather_lat";
    public static final String PREF_WEATHER_LON = "pref_weather_lon";
    
    private static final String WEATHER_OPTIONS_UPDATED = "pref_weather_options_updated";
    
    /**
     * Port number.
     */
    private boolean isCelsius;

    /**
     * Dark Sky key.
     */
    private String key;

    /**
     * Longitude.
     */
    private String longitude;

    /**
     * Latitude.
     */
    private String latitude;

    /**
     * Units.
     */
    private String units;

    /**
     * Preferences.
     */
    private final DPreference sharedPreferences;
    
    private DarkSkyOptions(DPreference sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    public boolean isValid() {
        return !TextUtils.isEmpty(latitude) &&
                !TextUtils.isEmpty(longitude) &&
                !TextUtils.isEmpty(units) &&
                !TextUtils.isEmpty(key);
    }
    
    /**
     * Construct a MqttOptions object from Configuration.
     */
    public static DarkSkyOptions from(DPreference sharedPreferences) {
        try {
            DarkSkyOptions options = new DarkSkyOptions(sharedPreferences);
            options.key = sharedPreferences.getPrefString(PREF_DARK_SKY_KEY, null);
            options.latitude = sharedPreferences.getPrefString(PREF_WEATHER_LAT, null);
            options.longitude = sharedPreferences.getPrefString(PREF_WEATHER_LON, null);
            options.isCelsius = sharedPreferences.getPrefBoolean(MQTTOptions.PREF_TLS_CONNECTION, false);
            options.units = sharedPreferences.getPrefString(PREF_WEATHER_UNITS, DarkSkyRequest.UNITS_US);
            return options;
        } catch (Exception e) {
            throw new IllegalArgumentException("While processing weather options", e);
        }
    }

    public void setLon(String longitude) {
        this.sharedPreferences.setPrefString(PREF_WEATHER_LON, longitude);
    }

    public void setLat(String latitude) {
        this.sharedPreferences.setPrefString(PREF_WEATHER_LAT, latitude);
    }

    public String getLatitude() {
        return latitude;
    }

    public String getLongitude() {
        return longitude;
    }

    public boolean getIsCelsius() {
        return isCelsius;
    }

    public String getWeatherUnits() {
        return units;
    }

    public String getDarkSkyKey() {
        return key;
    }

    public void setIsCelsius(boolean isCelsius) {
        sharedPreferences.setPrefString(PREF_WEATHER_UNITS, isCelsius ? DarkSkyRequest.UNITS_SI : DarkSkyRequest.UNITS_US);
        setOptionsUpdated(true);
    }

    public void setDarkSkyKey(String value) {
        this.sharedPreferences.setPrefString(PREF_DARK_SKY_KEY, value);
        setOptionsUpdated(true);
    }

    private void setOptionsUpdated(boolean value) {
        this.sharedPreferences.setPrefBoolean(WEATHER_OPTIONS_UPDATED, value);
    }

    public boolean hasUpdates() {
        boolean updates = sharedPreferences.getPrefBoolean(WEATHER_OPTIONS_UPDATED, false);
        if(updates) {
            setOptionsUpdated(false);
        }
        return updates;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof DarkSkyOptions)) {
            return false;
        }
        DarkSkyOptions o = (DarkSkyOptions) obj;
        return TextUtils.equals(key , o.key)
                && TextUtils.equals(units, o.units)
                && TextUtils.equals(latitude, o.latitude)
                && TextUtils.equals(longitude, o.longitude)
                && isCelsius == o.isCelsius;
    }
}

File: app/src/main/java/com/thanksmister/iot/mqtt/alarmpanel/network/DarkSkyOptions.kt
package com.thanksmister.iot.mqtt.alarmpanel.network

import android.text.TextUtils

import dpreference.DPreference
import javax.inject.Inject

/**
 * For original implementation see https://github.com/androidthings/sensorhub-cloud-iot.
 */
class DarkSkyOptions @Inject
constructor(private val sharedPreferences: DPreference) {

    val isValid: Boolean
        get() = !TextUtils.isEmpty(latitude) &&
                !TextUtils.isEmpty(longitude) &&
                !TextUtils.isEmpty(weatherUnits) &&
                !TextUtils.isEmpty(darkSkyKey)

    var darkSkyKey: String?
        get() = sharedPreferences.getPrefString(PREF_DARK_SKY_KEY, null)
        set(value) {
            this.sharedPreferences.setPrefString(PREF_DARK_SKY_KEY, value)
            setOptionsUpdated(true)
        }

    var weatherUnits: String
        get() = sharedPreferences.getPrefString(PREF_WEATHER_UNITS, DarkSkyRequest.UNITS_US)
        set(value) = this.sharedPreferences.setPrefString(PREF_WEATHER_UNITS, value)

    var longitude: String
        get() = sharedPreferences.getPrefString(PREF_WEATHER_LON, null)
        set(value) = this.sharedPreferences.setPrefString(PREF_WEATHER_LON, value)

    var latitude: String
        get() = sharedPreferences.getPrefString(PREF_WEATHER_LAT, null)
        set(value) = this.sharedPreferences.setPrefString(PREF_WEATHER_LAT, value)

    fun isCelsius(): Boolean {
        return sharedPreferences.getPrefString(PREF_WEATHER_UNITS, DarkSkyRequest.UNITS_US) == DarkSkyRequest.UNITS_SI
    }

    fun setIsCelsius(isCelsius: Boolean) {
        sharedPreferences.setPrefString(PREF_WEATHER_UNITS, if (isCelsius) DarkSkyRequest.UNITS_SI else DarkSkyRequest.UNITS_US)
        setOptionsUpdated(true)
    }

    private fun setOptionsUpdated(value: Boolean) {
        this.sharedPreferences.setPrefBoolean(WEATHER_OPTIONS_UPDATED, value)
    }

    fun hasUpdates(): Boolean {
        val updates = sharedPreferences.getPrefBoolean(WEATHER_OPTIONS_UPDATED, false)
        if (updates) {
            setOptionsUpdated(false)
        }
        return updates
    }

    companion object {

        val PREF_WEATHER_UNITS = "pref_weather_units"
        val PREF_DARK_SKY_KEY = "pref_dark_sky_key"
        val PREF_WEATHER_LAT = "pref_weather_lat"
        val PREF_WEATHER_LON = "pref_weather_lon"

        private val WEATHER_OPTIONS_UPDATED = "pref_weather_options_updated"

        /**
         * Construct a MqttOptions object from Configuration.
         */
        /*fun from(sharedPreferences: DPreference): DarkSkyOptions {
            try {
                val options = DarkSkyOptions(sharedPreferences)
                options.key =
                options.latitude = sharedPreferences.getPrefString(PREF_WEATHER_LAT, null)
                options.longitude = sharedPreferences.getPrefString(PREF_WEATHER_LON, null)
                options.isCelsius = sharedPreferences.getPrefString(PREF_WEATHER_UNITS, DarkSkyRequest.UNITS_US) == DarkSkyRequest.UNITS_SI
                options.weatherUnits = sharedPreferences.getPrefString(PREF_WEATHER_UNITS, DarkSkyRequest.UNITS_US)
                return options
            } catch (e: Exception) {
                throw IllegalArgumentException("While processing weather options", e)
            }

        }*/
    }
}